README.md
# **IPK Project 2 ZETA - Packet sniffer**

Network analyzer for capturing and filtering packets on a specific network interface

## **Introduction**

This program is a network analyzer for capturing and filtering packets on a specific network interface. It can capture and display packets of the following protocols: ARP, ICMPv4, ICMPv6, IGMP, MLD, NDP, TCP and UDP. It can also filter packets by port number and protocol.

## **Libraries**

This program uses the following libraries:

### **Standard C libraries**
- *stdio.h* - for input/output
- *stdlib.h* - for standard library functions
- *string.h* - for string functions
- *stdbool.h* - for boolean type
- *ctype.h* - for character functions
- *time.h* - for time functions
- *signal.h* - for SIGINT handling

### **Libraries from libpcap**
- *pcap.h* - main library for capturing and processing packets
- *netinet/in.h* - for IPv4 and IPv6 addresses
- *netinet/if_ether.h* - for Ethernet addresses
- *netinet/ip.h* - for IPv4 header
- *netinet/ip6.h* - for IPv6 header
- *netinet/tcp.h* - for TCP header
- *netinet/udp.h* - for UDP header
- *arpa/inet.h* - for converting IP addresses to strings

## **Code structure**

The structure could be divided into four main parts:

- ### **Parsing arguments**
- ### **Opening interface**
- ### **Capturing packets**
- ### **Printing packets info**

### **Parsing arguments**

First the program parses command line arguments and stores them in the structure FlagsT. The structure contains the following fields:

```c
struct FlagsT {
    bool interface_defined;
    char *interface; // interface name
    bool port_defined;
    int port; // port > 0 && port < 65536
    bool tcp_defined;
    bool udp_defined;
    bool arp_defined;
    bool icmp4_defined;
    bool icmp6_defined;
    bool igmp_defined;
    bool mld_defined;
    bool num_defined;
    int num; // num > 0
    bool ndp_defined;
};
```

During the process it is also building a packet filter string that will be used to filter packets in the next step. All of this is done in the function parse_input: 

```c
void parse_input(int argc, char **argv, struct FlagsT *flags, char *protocol_filter);
```

### **Opening interface**

After parsing the arguments, the program opens the interface and starts capturing packets. Device IP address and netmask are obtained using pcap_lookupnet. Then the program opens the interface using pcap_open_live. After that it runs pcap_compile to compile the packet filter string and pcap_setfilter to set the filter. All of this is done in the function pcap_handle_ctor which returns a pointer to the constructed pcap handle.

Disclaimer: Most of the code in this function is taken from this website: https://vichargrave.github.io/programming/develop-a-packet-sniffer-with-libpcap/

```c
pcap_t *pcap_handle_ctor(char *interface, char *protocol_filter);
```

### **Capturing packets**

When the pcap handle is constructed, the program checks if the data link type is Ethernet and then starts capturing packets. It uses pcap_loop function to capture packets. The function pcap_loop takes a callback function as an argument. The callback function is called for each captured packet and is defined in the function handle_packet.

```c
int check_datalink_header(pcap_t *handle);
```

### **Printing packets info**

Inside the callback function handle_packet, the program prints the following information about the packet:

- timestamp
- source MAC address
- destination MAC address
- frame length
- source IP address
- destination IP address
- source port (if TCP or UDP and port is specified)
- destination port (if TCP or UDP and port is specified)
- formatted payload data

The function handle_packet is defined as follows:

```c
void handle_packet(u_char *args, const struct pcap_pkthdr *header, const u_char *packet);
```

Inside the function, the program first formats and displays timestamp from packet header, then does the same for MAC addresses and frame length. After that it checks if the packet is IPv4, IPv6 or ARP. Then it checks if the packet is TCP or UDP. Based on this it prints the source and destination IP addresses and ports. At the end it formats and prints the payload data.

```c
// Formats and prints timestamp
void display_timestamp(const struct pcap_pkthdr *header);
// Formats and prints MAC addresses
void display_mac_addrs(char *src_mac_addr, char *dest_mac_addr, struct ether_header *eth_header);
// Formats and prints payload data
void display_data(const struct pcap_pkthdr *header, const u_char *packet);
```

### **Other functions**

The program also contains the following functions:

```c
void show_interfaces(void); // Prints all available interfaces
void sigint_handler(int sig); // Handles SIGINT signal
```

## **Usage**
```
$ make
$ sudo ./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] [--ndp] {-n num}
```
## **Flags**
```
-i interface | --interface interface
```
Specifies the interface to capture packets from. If not specified or the program is executed without any flags, all available interfaces will be listed.
```
-p port [--tcp|-t] [--udp|-u]
```
Listens for packets on the specified port. Can't be used without specifying either --tcp/-t or --udp/-u or both.

### **What is TCP and UDP?**
TCP (Tranmission Control Protocol) is connection-oriented,reliable, and error-checked protocol that transfers stream of bytes on transport layer.

UDP (User Datagram Protocol) is connectionless, unreliable, and simple protocol that transfers datagrams on transport layer.

```
--arp
```
Filters ARP packets.

### **What is ARP?**

ARP (Address Resolution Protocol) is a protocol used to connect IP addresses with MAC addresses. It is used to find the MAC address of a device on a network when only its IP address is known.
```
--icmp4
```
Filters ICMPv4 packets.

### **What is ICMPv4?**
ICMPv4 (Internet Control Message Protocol) is a protocol used to send error messages and operational information between network devices.
```
--icmp6
```
Filters ICMPv6 packets.

### **What is ICMPv6?**
ICMPv6 analogous to ICMPv4, but for IPv6.
```
--igmp
```
Filters IGMP packets.

### **What is IGMP?**
IGMP (Internet Group Management Protocol) is a protocol that allows multiple devices to have the same IP address and receive the same data.
```
--mld
```
Filters MLD packets.

### **What is MLD?**
MLD (Multicast Listener Discovery) is a protocol that allows routers to discover multicast listeners on the network.
```
--ndp
```
Filters NDP packets.

### **What is NDP?**
NDP (Neighbor Discovery Protocol) is a protocol that help the communication between neighboring hosts in local network.

## **Execution examples with outputs**

### **Capture packet on interface enp0s3**
```
$ sudo ./ipk-sniffer -i enp0s3
```

Output format

```
2023-04-17T19:20:56.222+00:00
src MAC: 08:00:27:32:28:e0
dst MAC: 38:43:7d:a6:b4:07
frame length: 90 bytes
src IP: 192.168.0.249
dst IP: 34.123.33.186
src port: 60416
dst port: 443

0x0000: 38 43 7d a6 b4 07 08 00 27 32 28 e0 08 00 45 00  8C}.....'2(...E.
0x0010: 00 4c 0b 25 40 00 40 06 29 b1 c0 a8 00 f9 22 7b  .L.%@.@.)....."{
0x0020: 21 ba ec 00 01 bb 7f 09 21 4d b6 38 f4 f3 80 18  !.......!M.8....
0x0030: 01 f5 06 15 00 00 01 01 08 0a 94 1e d9 aa 81 0f  ................
0x0040: 70 7c 17 03 03 00 13 65 59 b5 cc dd bc 92 28 4e  p|.....eY.....(N
0x0050: 1a 89 df c7 92 da cf 91 e9 3d                    .........=
```

### **List all available interfaces**
```
$ sudo ./ipk-sniffer
```
or
```
$ sudo ./ipk-sniffer [-i | --interface]
```

Output format

```
Interface: enp0s3
Interface: any
Interface: lo
Interface: nflog
Interface: nfqueue
```

### **Capture 2 tcp packets on port 20 on interface enp0s3**
```
$ sudo ./ipk-sniffer -i enp0s3 -p 20 --tcp -n 2
```

Output format

```
2023-04-17T19:26:33.141+00:00
src MAC: 08:00:27:32:28:e0
dst MAC: 38:43:7d:a6:b4:07
frame length: 74 bytes
src IP: 192.168.0.249
dst IP: 20.189.173.10
src port: 33248
dst port: 443

0x0000: 38 43 7d a6 b4 07 08 00 27 32 28 e0 08 00 45 00  8C}.....'2(...E.
0x0010: 00 3c 2b d0 40 00 40 06 8b 83 c0 a8 00 f9 14 bd  .<+.@.@.........
0x0020: ad 0a 81 e0 01 bb 29 2f 63 06 00 00 00 00 a0 02  ......)/c.......
0x0030: fa f0 83 97 00 00 02 04 05 b4 04 02 08 0a b0 8b  ................
0x0040: f3 7b 00 00 00 00 01 03 03 07                    .{........

2023-04-17T19:26:33.380+00:00
src MAC: 38:43:7d:a6:b4:07
dst MAC: 08:00:27:32:28:e0
frame length: 66 bytes
src IP: 20.189.173.10
dst IP: 192.168.0.249
src port: 443
dst port: 33248

0x0000: 08 00 27 32 28 e0 38 43 7d a6 b4 07 08 00 45 00  ..'2(.8C}.....E.
0x0010: 00 34 2f 35 40 00 68 06 60 26 14 bd ad 0a c0 a8  .4/5@.h.`&......
0x0020: 00 f9 01 bb 81 e0 00 a2 e8 45 29 2f 63 07 80 12  .........E)/c...
0x0030: ff ff f3 05 00 00 02 04 05 8c 01 03 03 08 01 01  ................
0x0040: 04 02                                            ..
```

### **Capture arp or icmp4 or ndp packet on interface enp0s3**
```
$ sudo ./ipk-sniffer -i enp0s3 --arp --icmp4 --ndp
```

Output format

```
2023-04-17T19:28:57.068+00:00
src MAC: 38:43:7d:a6:b4:07
dst MAC: ff:ff:ff:ff:ff:ff
frame length: 56 bytes
src IP: 192.168.0.1
dst IP: 192.168.0.101

0x0000: ff ff ff ff ff ff 38 43 7d a6 b4 07 08 06 00 01  ......8C}.......
0x0010: 08 00 06 04 00 01 38 43 7d a6 b4 07 c0 a8 00 01  ......8C}.......
0x0020: 00 00 00 00 00 00 c0 a8 00 65 00 00 00 00 00 00  .........e......
0x0030: 00 00 00 00 00 00 00 00                          ........
```

## **Testing**

Testing was done manually using application with similar functionality, Wireshark.
During testing, I captured packets from my home network and compared them with the output of my application. I also tried to raise the number of captured packets to 1000 and it worked without any problems and to identify as many edge cases as possible so the program would not crash without proper error message. 

## **Conclusion**

The application is able to capture defined amount of packets from specified available interface and filter them by protocol and port number. It is also able to list all available interfaces. I learned a lot about network protocols, packets and how to work with them. Even though it was a lot of work, I enjoyed it and I am satisfied with the result.

## **References**

- TCP - https://cs.wikipedia.org/wiki/Transmission_Control_Protocol

- UDP - https://www.techtarget.com/searchnetworking/definition/UDP-User-Datagram-Protocol

- ARP - https://www.javatpoint.com/address-resolution-protocol-and-its-types

- ICMP - https://www.fortinet.com/resources/cyberglossary/internet-control-message-protocol-icmp

- IGMP - https://www.cloudflare.com/learning/network-layer/what-is-igmp/

- MLD - https://kb.netgear.com/21991/What-is-Multicast-Listener-Discovery-MLD-and-how-does-it-work-with-my-managed-switch

- NDP - https://www.ionos.com/digitalguide/server/know-how/what-is-neighborhood-discovery-protocolndp/

- Opening interface - https://vichargrave.github.io/programming/develop-a-packet-sniffer-with-libpcap/

CHANGELOG.md
# Changelog

## Version 0.1.0
- repository init
- files created

## Version 0.2.0
- added command line arguments parser

## Version 0.3.0
- added packet handle creator

## Version 0.4.0
- restriction of port numbers
- added filter string for packet filtering

## Version 0.5.0
- added support for Ethernet data link type

## Version 0.6.0
- fixed return types
- added packet timestamps

## Version 1.0.0
- display MAC addresses
- display IP addresses
- display payload data
- added sigint handling

## Version 1.1.0
- [header file](ipk-sniffer.h)
- refactoring

## Version 1.2.0
- added [LICENSE](LICENSE)
- added [CHANGELOG.md](CHANGELOG.md)

## Version 1.3.0
- finished [README.md](README.md)
- sigint handler hotfix

ipk-sniffer.h
// FILE: ipk-sniffer.h
// AUTHOR: xzavad20
// DESC: Network analyzer, captures and filters packets on a specific network interface

#ifndef __SNIFFER_H__
#define __SNIFFER_H__

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>
#include <pcap.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <time.h>
#include <netinet/ip.h>
#include <netinet/if_ether.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/ip6.h>
#include <signal.h>

#define OK 0
#define ERROR 1
#define MAC_ADDR_LEN 18
#define LINE_BYTES 16

// Struct for command line arguments
struct FlagsT {
    bool interface_defined;
    char *interface;
    bool port_defined;
    int port;
    bool tcp_defined;
    bool udp_defined;
    bool arp_defined;
    bool icmp4_defined;
    bool icmp6_defined;
    bool igmp_defined;
    bool mld_defined;
    bool num_defined;
    int num;
    bool ndp_defined;
};

pcap_t *pcap_handle = NULL;

// Function prototypes

// Handler for SIGINT
void sigint_handler(int sig);

// Displays all available interfaces
void show_interfaces(void);

// Parses command line arguments
void parse_input(int argc, char **argv, struct FlagsT *flags, char *protocol_filter);

// Displays packet timestamp
void display_timestamp(const struct pcap_pkthdr *header);

// Displays packet MAC addresses
void display_mac_addrs(char *src_mac_addr, char *dest_mac_addr, struct ether_header *eth_header);

// Displays packet payload data
void display_data(const struct pcap_pkthdr *header, const u_char *packet);

// Process packet and display its data
void handle_packet(u_char *args, const struct pcap_pkthdr *header, const u_char *packet);

// Opens interface for packet capturing and filtering
pcap_t *pcap_handle_ctor(char *interface, char *protocol_filter);

// Checks if data link type is Ethernet
int check_datalink_header(pcap_t *handle);

#endif

ipk-sniffer.c
// FILE: ipk-sniffer.c
// AUTHOR: xzavad20
// DESC: Network analyzer, captures and filters packets on a specific network interface

#include "ipk-sniffer.h"

void sigint_handler(int sig) {
    fprintf(stderr, "Interrupted by SIGINT, gracefully exiting...\n");
    if (pcap_handle != NULL) {
        pcap_close(pcap_handle);
    }
    exit(sig);
}

void show_interfaces(void) {
    pcap_if_t *interface = NULL;
    char errbuf[PCAP_ERRBUF_SIZE] = {0};
    if (pcap_findalldevs(&interface, errbuf) != 0) {
        fprintf(stderr, "pcap_findalldev: Interfaces display error %s\n", errbuf);
        exit(ERROR);
    }
    else {
        // Display available interfaces
        while(interface != NULL) {
            printf("Interface: %s\n", interface->name);
            interface = interface->next;
        }
        pcap_freealldevs(interface);
    }
}

void parse_input(int argc, char **argv, struct FlagsT *flags, char *protocol_filter) {
    memset(flags, 0, sizeof(struct FlagsT));
    if (argc == 1) {
        show_interfaces();
        exit(OK);
    }
    for (int i = 1; i < argc; i++) {
        if ((strcmp(argv[i], "-i") == 0) || (strcmp(argv[i], "--interface") == 0)) {
            if (argc == 2 && i+1 >= argc) {
                show_interfaces();
                exit(OK);
            }
            if (++i < argc) {
                if (flags->interface_defined) {
                    fprintf(stderr, "Error: interface specified multiple times\n");
                    exit(ERROR);
                }
                flags->interface_defined = true;
                flags->interface = malloc(sizeof(char) * (strlen(argv[i]) + 1));
                if (flags->interface == NULL) {
                    fprintf(stderr, "Error: failed to allocate memory\n");
                    exit(ERROR);
                }
                strcpy(flags->interface, argv[i]);
            } 
            else {
                fprintf(stderr, "Error: missing interface name\n");
                exit(ERROR);
            }
        } 
        else if (strcmp(argv[i], "-p") == 0 || strcmp(argv[i], "--port") == 0) {
            if (flags->port_defined) {
                fprintf(stderr, "Error: port specified multiple times\n");
                exit(ERROR);
            }
            if (++i < argc) {
                char *endptr;
                long int port_num = strtol(argv[i], &endptr, 10);
                if (*endptr != '\0' || port_num < 0 || port_num > 65535) {
                    fprintf(stderr, "Error: invalid port number '%s'\n", argv[i]);
                    exit(ERROR);
                } 
                else {
                    flags->port = (int)port_num;
                }
                flags->port_defined = true;
            } 
            else {
                fprintf(stderr, "Error: missing port number\n");
                exit(ERROR);
            }
        } 
        else if (strcmp(argv[i], "--tcp") == 0 || strcmp(argv[i], "-t") == 0) { 
            if (flags->tcp_defined) {
                fprintf(stderr, "Error: only one of -t or --tcp may be specified\n");
                exit(ERROR);
            }
            flags->tcp_defined = true; 
        }
        else if (strcmp(argv[i], "--udp") == 0 || strcmp(argv[i], "-u") == 0) { 
            if (flags->udp_defined) {
                fprintf(stderr, "Error: only one of -u or --udp may be specified\n");
                exit(ERROR);
            }
            flags->udp_defined = true; 
        }
        else if (strcmp(argv[i], "--arp") == 0) {
            if (flags->arp_defined) {
                fprintf(stderr, "Error: only one of -a or --arp may be specified\n");
                exit(ERROR);
            }
            flags->arp_defined = true;
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "arp ");
            }
            else {
                strcat(protocol_filter, "or arp ");
            }
        }
        else if (strcmp(argv[i], "--icmp4") == 0) { 
            if (flags->icmp4_defined) {
                fprintf(stderr, "Error: only one of -i or --icmp4 may be specified\n");
                exit(ERROR);
            }
            flags->icmp4_defined = true;
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "icmp ");
            }
            else {
                strcat(protocol_filter, "or icmp ");
            }
        }
        else if (strcmp(argv[i], "--icmp6") == 0) { 
            if (flags->icmp6_defined) {
                fprintf(stderr, "Error: only one of -i or --icmp6 may be specified\n");
                exit(ERROR);
            }
            flags->icmp6_defined = true; 
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "(icmp6[0] == 128 or icmp6[0] == 129) ");
            }
            else {
                strcat(protocol_filter, "or (icmp6[0] == 128 or icmp6[0] == 129) ");
            }
        }
        else if (strcmp(argv[i], "--igmp") == 0) { 
            if (flags->igmp_defined) {
                fprintf(stderr, "Error: only one of -i or --igmp may be specified\n");
                exit(ERROR);
            }
            flags->igmp_defined = true; 
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "igmp ");
            }
            else {
                strcat(protocol_filter, "or igmp ");
            }
        }
        else if (strcmp(argv[i], "--mld") == 0) {
            if (flags->mld_defined) {
                fprintf(stderr, "Error: only one of -i or --mld may be specified\n");
                exit(ERROR);
            }
            flags->mld_defined = true;
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "(icmp6[0] == 130 or icmp6[0] == 131 or icmp6[0] == 132 or icmp6[0] == 143) ");
            }
            else {
                strcat(protocol_filter, "or (icmp6[0] == 130 or icmp6[0] == 131 or icmp6[0] == 132 or icmp6[0] == 143) ");
            }
        }
        else if (strcmp(argv[i], "--ndp") == 0) {
            if (flags->ndp_defined) {
                fprintf(stderr, "Error: only one of -i or --ndp may be specified\n");
                exit(ERROR);
            }
            flags->ndp_defined = true;
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "(icmp6[0] == 133 or icmp6[0] == 134 or icmp6[0] == 135 or icmp6[0] == 136 or icmp6[0] == 137) ");
            }
            else {
                strcat(protocol_filter, "or (icmp6[0] == 133 or icmp6[0] == 134 or icmp6[0] == 135 or icmp6[0] == 136 or icmp6[0] == 137) ");
            }
        }
        else if (strcmp(argv[i], "-n") == 0) {
            if (flags->num_defined) {
                fprintf(stderr, "Error: only one of -n or --num may be specified\n");
                exit(ERROR);
            }
            if (++i < argc) {
                char *endptr;
                long int num_val = strtol(argv[i], &endptr, 10);
                if (*endptr != '\0') {
                    fprintf(stderr, "Error: invalid number '%s'\n", argv[i]);
                    exit(ERROR);
                } 
                else {
                    if (num_val > 0) {
                        flags->num = (int)num_val;
                    }
                    else {
                        fprintf(stderr, "Number has to be greater than 0, received '%ld'\n", num_val);
                        exit(ERROR);
                    }
                    flags->num_defined = true;
                }
            } 
            else {
                fprintf(stderr, "Error: missing number\n");
                exit(ERROR);
            }
        } 
        else {
            fprintf(stderr, "Error: unknown flag '%s'\n", argv[i]);
            exit(ERROR);
        }
    }
    if (flags->port_defined == false) {
        if (flags->tcp_defined == true) {
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "tcp ");
            }
            else {
                strcat(protocol_filter, "or tcp ");
            }
        }
        if (flags->udp_defined == true) {
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "udp ");
            }
            else {
                strcat(protocol_filter, "or udp ");
            }
        }
    }
    else {
        char port_temp[7] = ""; 
        if (flags->tcp_defined == false && flags->udp_defined == false) {
            fprintf(stderr, "Error: When port is defined there must be either -t or -u\n");
            exit(ERROR);
        }
        if (flags->tcp_defined == true) {
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "(tcp port ");
                sprintf(port_temp, "%d) ", flags->port);
                strcat(protocol_filter, port_temp);
            }
            else {
                strcat(protocol_filter, "or (tcp port ");
                sprintf(port_temp, "%d) ", flags->port);
                strcat(protocol_filter, port_temp);
            }
        }
        if (flags->udp_defined == true) {
            if (strlen(protocol_filter) == 0) {
                strcat(protocol_filter, "(udp port ");
                sprintf(port_temp, "%d) ", flags->port);
                strcat(protocol_filter, port_temp);
            }
            else {
                strcat(protocol_filter, "or (udp port ");
                sprintf(port_temp, "%d) ", flags->port);
                strcat(protocol_filter, port_temp);
            }
        }
    }
    if (flags->tcp_defined == false && flags->udp_defined == false && flags->arp_defined == false && flags->icmp4_defined == false && flags->icmp6_defined == false && flags->igmp_defined == false && flags->mld_defined == false && flags->ndp_defined == false) {
        strcat(protocol_filter, "tcp or udp or arp or icmp or (icmp6[0] == 128 or icmp6[0] == 129) or igmp or (icmp6[0] == 130 or icmp6[0] == 131 or icmp6[0] == 132 or icmp6[0] == 143) or (icmp6[0] == 133 or icmp6[0] == 134 or icmp6[0] == 135 or icmp6[0] == 136 or icmp6[0] == 137)");
    }
    if (flags->num_defined == false) {
        flags->num = 1;
    }
}

void display_timestamp(const struct pcap_pkthdr *header) {
    // Timestamp from Ethernet packet header
    time_t timestamp = header->ts.tv_sec;
    // Convert timestamp to local time
    struct tm *tm_local = localtime(&timestamp);
    char formatted_ts[32];
    // Format timestamp to yyyy-mm-ddThh:mm:ss
    strftime(formatted_ts, 32, "%Y-%m-%dT%H:%M:%S", tm_local);
    sprintf(&formatted_ts[19], ".%03d", (int) header->ts.tv_usec / 1000);

    // Calculate time zone offset
    int tz_offset_hours = (tm_local->tm_gmtoff / 3600) % 24;
    int tz_offset_minutes = (tm_local->tm_gmtoff / 60) % 60;
    char tz_sign = (tm_local->tm_gmtoff >= 0) ? '+' : '-';
    // Format time zone to +-hh:mm
    sprintf(&formatted_ts[23], "%c%02d:%02d", tz_sign, tz_offset_hours, tz_offset_minutes);

    printf("%s\n", formatted_ts);
}

void display_mac_addrs(char *src_mac_addr, char *dest_mac_addr, struct ether_header *eth_header) {
    // Get MAC adresses from header
    snprintf(src_mac_addr, MAC_ADDR_LEN, "%02x:%02x:%02x:%02x:%02x:%02x", eth_header->ether_shost[0], 
        eth_header->ether_shost[1], eth_header->ether_shost[2], eth_header->ether_shost[3], eth_header->ether_shost[4], eth_header->ether_shost[5]);
    snprintf(dest_mac_addr, MAC_ADDR_LEN, "%02x:%02x:%02x:%02x:%02x:%02x", eth_header->ether_dhost[0],
        eth_header->ether_dhost[1], eth_header->ether_dhost[2], eth_header->ether_dhost[3], eth_header->ether_dhost[4], eth_header->ether_dhost[5]);
    printf("src MAC: %s\n", src_mac_addr);
    printf("dst MAC: %s\n", dest_mac_addr);
}

void display_ip_addrs(const void *restrict src, const void *restrict dest) {
    printf("src IP: %s\n", (char *)src);
    printf("dst IP: %s\n", (char *)dest);
}

void display_data(const struct pcap_pkthdr *header, const u_char *packet) {
    unsigned int byte_offset = 0;
    int bytes_left_print, bytes_left;

    while (byte_offset < header->len) {
        bytes_left = header->len - byte_offset;
        if (bytes_left < LINE_BYTES) {
            bytes_left_print = bytes_left;
        }
        else {
            bytes_left_print = LINE_BYTES;
        }
        printf("0x%04x: ", byte_offset);
        for (int i = 0; i < bytes_left_print; i++) {
            printf("%02x ", packet[byte_offset + i]);
        }
        for (int j = 0; j < LINE_BYTES - bytes_left_print; j++) {
            printf("   ");
        }
        printf(" ");
        for (int k = 0; k < bytes_left_print; k++) {
            char c = packet[byte_offset + k];
            // Display printable characters
            if (isprint(c)) {
                printf("%c", c);
            } 
            else {
                printf(".");
            }
        }
        printf("\n");
        byte_offset += bytes_left_print;
    }
    printf("\n");
}

void handle_packet(u_char *args, const struct pcap_pkthdr *header, const u_char *packet) {
    char src_mac_addr[MAC_ADDR_LEN], dest_mac_addr[MAC_ADDR_LEN] = "";
    char src_ip_addr[INET_ADDRSTRLEN], dest_ip_addr[INET_ADDRSTRLEN] = "";
    struct ether_header *eth_header;
    eth_header = (struct ether_header *)packet;

    signal(SIGINT, sigint_handler);

    // Display timestamp
    display_timestamp(header);

    // Display mac addresses
    display_mac_addrs(src_mac_addr, dest_mac_addr, eth_header);

    // Display frame length
    printf("frame length: %d bytes\n", header->len);

    // Format the output based on protocol (IPv4 or IPv6 or ARP)
    if (ntohs(eth_header->ether_type) == ETHERTYPE_IP) {
        struct ip *ipv4_header = (struct ip *)(packet + sizeof(struct ether_header));
        inet_ntop(AF_INET, &ipv4_header->ip_src, src_ip_addr, INET_ADDRSTRLEN);
        inet_ntop(AF_INET, &ipv4_header->ip_dst, dest_ip_addr, INET_ADDRSTRLEN);
        display_ip_addrs(src_ip_addr, dest_ip_addr);
        if (ipv4_header->ip_p == IPPROTO_TCP) {
            struct tcphdr *tcp_header = (struct tcphdr *)(packet + sizeof(struct ether_header) + sizeof(struct ip));
            printf("src port: %d\n", ntohs(tcp_header->th_sport));
            printf("dst port: %d\n", ntohs(tcp_header->th_dport));
        } 
        else if (ipv4_header->ip_p == IPPROTO_UDP) {
            struct udphdr *udp_header = (struct udphdr *)(packet + sizeof(struct ether_header) + sizeof(struct ip));
            printf("src port: %d\n", ntohs(udp_header->uh_sport));
            printf("dst port: %d\n", ntohs(udp_header->uh_dport));
        }
    }
    else if (ntohs(eth_header->ether_type) == ETHERTYPE_IPV6) {
        struct ip6_hdr *ipv6_header = (struct ip6_hdr *)(packet + sizeof(struct ether_header));
        inet_ntop(AF_INET6, &ipv6_header->ip6_src, src_ip_addr, INET6_ADDRSTRLEN);
        inet_ntop(AF_INET6, &ipv6_header->ip6_dst, dest_ip_addr, INET6_ADDRSTRLEN);
        display_ip_addrs(src_ip_addr, dest_ip_addr);
        if (ipv6_header->ip6_nxt == IPPROTO_TCP) {
            struct tcphdr *tcp_header = (struct tcphdr *)(packet + sizeof(struct ether_header) + sizeof(struct ip6_hdr));
            printf("src port: %d\n", ntohs(tcp_header->th_sport));
            printf("dst port: %d\n", ntohs(tcp_header->th_dport));
        } 
        else if (ipv6_header->ip6_nxt == IPPROTO_UDP) {
            struct udphdr *udp_header = (struct udphdr *)(packet + sizeof(struct ether_header) + sizeof(struct ip6_hdr));
            printf("src port: %d\n", ntohs(udp_header->uh_sport));
            printf("dst port: %d\n", ntohs(udp_header->uh_dport));
        }
    }
    else if (ntohs(eth_header->ether_type) == ETHERTYPE_ARP) {
        struct ether_arp *arp_header = (struct ether_arp *)(packet + sizeof(struct ether_header));
        inet_ntop(AF_INET, &arp_header->arp_spa, src_ip_addr, INET_ADDRSTRLEN);
        inet_ntop(AF_INET, &arp_header->arp_tpa, dest_ip_addr, INET_ADDRSTRLEN);
        display_ip_addrs(src_ip_addr, dest_ip_addr);
    }
    printf("\n");

    // Display payload data
    display_data(header, packet);
}

// SOURCE: https://vichargrave.github.io/programming/develop-a-packet-sniffer-with-libpcap/
pcap_t *pcap_handle_ctor(char *interface, char *protocol_filter) {
    char errbuf[PCAP_ERRBUF_SIZE];
    pcap_t *pcap_handle = NULL;
    struct bpf_program bpf;
    bpf_u_int32 netmask;
    bpf_u_int32 srcip;

    // Get device source ip address and netmask
    if (pcap_lookupnet(interface, &srcip, &netmask, errbuf) == PCAP_ERROR) {
        fprintf(stderr, "pcap_lookupnet: %s\n", errbuf);
        return NULL;
    }

    // Open the device for packet capturing
    pcap_handle = pcap_open_live(interface, BUFSIZ, 1, 1000, errbuf);
    if (pcap_handle == NULL) {
        fprintf(stderr, "pcap_open_live(): %s\n", errbuf);
        return NULL;
    }

    // Convert filter expression to binary
    if (pcap_compile(pcap_handle, &bpf, protocol_filter, 0, netmask) == PCAP_ERROR) {
        fprintf(stderr, "pcap_compile(): %s\n", pcap_geterr(pcap_handle));
        return NULL;
    }

    // Bind the packet filter to the libpcap handle.
    if (pcap_setfilter(pcap_handle, &bpf) == PCAP_ERROR) {
        fprintf(stderr, "Error: pcap_setfilter() ... %s\n", pcap_geterr(pcap_handle));
        return NULL;
    }
    return pcap_handle;
}

int check_datalink_header(pcap_t *pcap_handle) {
    int datalink_type;
    if ((datalink_type = pcap_datalink(pcap_handle)) == PCAP_ERROR) {
        fprintf(stderr, "pcap_datalink(): Function execution error%s\n", pcap_geterr(pcap_handle));
        return ERROR;
    }
    if (datalink_type != DLT_EN10MB) {
        fprintf(stderr, "Unsupported data link type %d, needs to be Ethernet\n", datalink_type);
        return ERROR;
    }
    return OK;
}

int main(int argc, char **argv) {
    struct FlagsT flags;
    char protocol_filter[512] = "";
    pcap_t *pcap_handle;
    int datalink_type;
    parse_input(argc, argv, &flags, protocol_filter);
    pcap_handle = pcap_handle_ctor(flags.interface, protocol_filter);
    if (pcap_handle == NULL) {
        fprintf(stderr, "Error: pcap_handle_ctor() ... Function execution error\n");
        exit(ERROR);
    }
    signal(SIGINT, sigint_handler);
    if ((datalink_type = check_datalink_header(pcap_handle)) == ERROR) {
        exit(ERROR);
    }
    if (pcap_loop(pcap_handle, flags.num, handle_packet, NULL) != 0) {
        fprintf(stderr, "Error: pcap_loop() ... Function execution error\n");
        exit(ERROR);
    }
    pcap_close(pcap_handle);
    return OK;   
}